### PR TITLE
Make RemoteData be friendly with Android Parcelable framework

### PR DESCRIPTION
**Pull request by KeithYokoma**
## Objective

Let `RemoteData` implement `Parcelable` to make it easy to use in the Android framework.

## Description

`RemoteData` is very useful to describe a state of asynchronous tasks like API calls, but it cannot be saved into `android.os.Bundle` as it's not `Parcelable` nor `Serializable`, thus `RemoteData` field in the Activity/Fragment cannot be saved on configuration changes and/or activity restoration.

AAC ViewModel with SavedStateHandle also requires a data compatible with `android.os.Bundle` so it's more and more important to have `RemoteData` implement `Parcelable`.

There are a few difference from the current `RemoteData` for pure Kotlin implementation:
1. Failure cannot hold `Exception`, but have `ErrorKind` interface instead for Parcelable Creator constraints.
2. Success value type is annotated with `@RawValue` of Parcelize feature, allowing value type to be any type that is compatible with `Parcelable` framework.
  This has a drawback that we cannot have a compile-time type check that `RemoteData.Success` value type needs to be one of the types being compatible with Parcelable, and it runs into an error at runtime when we pass a non-Parcelable value.

## Testing

`remotedata_android` module has 2 types of unit-test.
1. Copy of `RemoteDataTest` from `remotedata` module, to check if the higher order functions behaves in the same way as `RemoteData` in the `remotedata` module.
2. Robolectric tests to check `Parcelable` behavior, by marshalling a `RemoteData` object to the ByteArray and unmarshalling from the ByteArray, comparing the original one and unmarshalled one.

## TODOs

- [x] Add artifact deploy task to `remotedata_android/build.gradle.kts`
- [x] Configure CircleCI to build android artifact

### CODE REVIEW

**Review by yhanada — COMMENTED** [review]
This PR is still a draft, so some modifications may still be added, but at this point it looks good to me. 👍

**Comment by kittinunf** [comment]
I think this is a great great addition to the project. Thanks @KeithYokoma 👏 👏 👏

File: buildSrc/src/main/kotlin/Config.kt [hunk 1]
```diff
@@ -10,15 +10,21 @@ object MavenUrl {
...
 
 object Version {
+  const val android = "3.6.2"
+  const val androidJunit5 = "1.6.0.0"
   const val bintray = "1.8.4"
   const val kotlin = "1.3.30"
   const val kluent = "1.60"
   const val spek = "2.0.10"
   const val junitPlatform = "1.2.0"
+  const val junitEngine = "5.6.0"
```
Comment (kittinunf) [inline buildSrc/src/main/kotlin/Config.kt:20]: Bleeding edge stuff? Lol

Comment (KeithYokoma) [reply buildSrc/src/main/kotlin/Config.kt:20]: haha, it's the same as Spek 2.0.10 :)

File: remotedata_android/src/test/java/com/mercari/remotedata/android/ParcelableRemoteDataTest.kt [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import org.amshove.kluent.shouldBeEqualTo
+import org.amshove.kluent.shouldBeInstanceOf
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import java.lang.IllegalStateException
+import java.lang.RuntimeException
+
+@RunWith(RobolectricTestRunner::class)
+class ParcelableRemoteDataTest {
```
Comment (kittinunf) [inline remotedata_android/src/test/java/com/mercari/remotedata/android/ParcelableRemoteDataTest.kt:15]: 👏

File: remotedata_android/src/test/java/com/mercari/remotedata/android/ParcelableRemoteDataTest.kt [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        val marshalled: ByteArray = marshall(value)
+
+        // When: Store it to the Bundle
+        val restored: RemoteData.Failure<ErrorKind> = unmarshall(marshalled)
+
+        // Then: Restored value should be the same as the initial value
+        restored.error.shouldBeInstanceOf<NetworkError>()
+    }
+
+    @Test(expected = RuntimeException::class)
```
Comment (kittinunf) [inline remotedata_android/src/test/java/com/mercari/remotedata/android/ParcelableRemoteDataTest.kt:120]: [super nits] I think this is okay by itself, but I find the `assertThrows` to be much simpler to use.

**Review by yhanada — APPROVED** [review]

File: remotedata_android/src/main/java/com/mercari/remotedata/android/RemoteData.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        val isIndeterminateProgress: Boolean = progress == null
+
+        override fun equals(other: Any?): Boolean =
+                if (other === this) true
+                else {
+                    other is Loading<*> && other.progress == progress && other.totalUnits == totalUnits
+                }
+
+        override fun hashCode(): Int =
```
Comment (hotchemi) [inline remotedata_android/src/main/java/com/mercari/remotedata/android/RemoteData.kt:58]: nits: perhaps we can use `Objects.hash`?

**Review by hotchemi — APPROVED** [review]
Looks good to me👍

**Pull request merged**